File: dotnet/Varar.Core/Diagnostics.cs
using System.Collections.Immutable;

namespace Varar.Core;

public enum Severity
{
    Error,
    Warning,
}

public enum DiagnosticCode
{
    AmbiguousMatch,
    ErrorFenceWithoutStep,
    Drift,

    /// <summary>
    /// Reference blocks (ADR 0016): a link that resolves to no oath, to a section with no steps,
    /// or to a chain that reaches itself.
    /// </summary>
    ReferenceNotFound,
    ReferenceEmpty,
    ReferenceCycle,
}

/// <summary>A diagnostic on the shared rail. Port of <c>diagnostics.ts</c>.</summary>
public sealed record Diagnostic(Severity Severity, DiagnosticCode Code, string Message, Span Span);

public sealed record Candidate(string Expression, string SourceFile, int SourceLine);

public static class Diagnostics
{
    public static string ToWire(this Severity severity) => severity switch
    {
        Severity.Error => "error",
        Severity.Warning => "warning",
        _ => throw new ArgumentOutOfRangeException(nameof(severity), severity, null),
    };

    public static string ToWire(this DiagnosticCode code) => code switch
    {
        DiagnosticCode.AmbiguousMatch => "ambiguous-match",
        DiagnosticCode.ErrorFenceWithoutStep => "error-fence-without-step",
        DiagnosticCode.Drift => "drift",
        DiagnosticCode.ReferenceNotFound => "reference-not-found",
        DiagnosticCode.ReferenceEmpty => "reference-empty",
        DiagnosticCode.ReferenceCycle => "reference-cycle",
        _ => throw new ArgumentOutOfRangeException(nameof(code), code, null),
    };

    public static Diagnostic AmbiguousMatch(string text, Span span, ImmutableArray<Candidate> candidates)
    {
        var lines = string.Join("\n", candidates.Select(c => $"  '{c.Expression}'    at {c.SourceFile}:{c.SourceLine}"));
        return new Diagnostic(
            Severity.Error,
            DiagnosticCode.AmbiguousMatch,
            $"Ambiguous step: \"{text}\"\nMatched by:\n{lines}",
            span);
    }

    public static Diagnostic DriftDetected(string name, Span span) => new(
        Severity.Error,
        DiagnosticCode.Drift,
        $"This paragraph was an example and no longer matches any step (drift): \"{name}\".\n" +
        "Fix the step so it matches again, or accept it as prose (run in update mode).",
        span);

    public static Diagnostic ErrorFenceWithoutStep(Span span) => new(
        Severity.Error,
        DiagnosticCode.ErrorFenceWithoutStep,
        "This `error` fence marks the example as expected-to-fail, but the example has no step to run.",
        span);

    /// <summary>
    /// A reference block (ADR 0016) points at an oath the workspace does not hold. Never prose: a
    /// link-only block that resolves to nothing has no other reading, so it fails the run rather
    /// than degrading silently.
    /// </summary>
    public static Diagnostic ReferenceNotFound(string text, string path, Span span) => new(
        Severity.Error,
        DiagnosticCode.ReferenceNotFound,
        $"Reference to \"{text}\" points at \"{path}\", which is not an oath in this workspace.\n" +
        "Check the path, and that the file is matched by the `docs` globs in varar.config.json.",
        span);

    /// <summary>
    /// The referenced document exists but the section contributes no steps — a mistyped anchor, or
    /// a section that is pure prose.
    /// </summary>
    public static Diagnostic ReferenceEmpty(string text, string path, string slug, Span span) => new(
        Severity.Error,
        DiagnosticCode.ReferenceEmpty,
        $"Reference to \"{text}\" resolves to \"{(slug.Length == 0 ? path : $"{path}#{slug}")}\", " +
        "which contributes no steps.\nCheck the heading the anchor names, and that its section " +
        "contains a matching paragraph.",
        span);

    /// <summary>
    /// References nest to any depth, so a chain that reaches a section already on it is reported
    /// rather than recursed into.
    /// </summary>
    public static Diagnostic ReferenceCycle(IEnumerable<string> chain, Span span) => new(
        Severity.Error,
        DiagnosticCode.ReferenceCycle,
        $"Reference cycle: {string.Join(" \u2192 ", chain)}.",
        span);
}

File: typescript/packages/core/src/diagnostics.ts
import type { Span } from './span.ts'

export type Severity = 'error' | 'warning'

export type Diagnostic = {
  readonly severity: Severity
  readonly code: DiagnosticCode
  readonly message: string
  readonly span: Span
}

export type DiagnosticCode =
  | 'ambiguous-match'
  | 'error-fence-without-step'
  | 'drift'
  | 'reference-not-found'
  | 'reference-empty'
  | 'reference-cycle'

export type Candidate = {
  readonly expression: string
  readonly sourceFile: string
  readonly sourceLine: number
}

export type AmbiguousInput = {
  readonly text: string
  readonly span: Span
  readonly candidates: ReadonlyArray<Candidate>
}

export function ambiguousMatch(input: AmbiguousInput): Diagnostic {
  const lines = input.candidates
    .map((c) => `  '${c.expression}'    at ${c.sourceFile}:${c.sourceLine}`)
    .join('\n')
  return {
    severity: 'error',
    code: 'ambiguous-match',
    message: `Ambiguous step: "${input.text}"\nMatched by:\n${lines}`,
    span: input.span,
  }
}

// A paragraph the baseline recorded as an example no longer matches any step:
// drift. Rides the shared Diagnostic rail so every surface reports it the same
// way — a non-zero CLI exit, a failing vitest/pytest test, an LSP squiggle.
// `span` points at the drifted paragraph. Cleared by accepting (update mode).
export function driftDetected(input: { readonly name: string; readonly span: Span }): Diagnostic {
  return {
    severity: 'error',
    code: 'drift',
    message:
      `This paragraph was an example and no longer matches any step (drift): "${input.name}".\n` +
      'Fix the step so it matches again, or accept it as prose (run in update mode).',
    span: input.span,
  }
}

// An `error` fence declares its example expected-to-fail, but the example has
// no runnable step to produce that failure (nothing matched, or the match was
// ambiguous). `span` points at the orphaned fence.
export function errorFenceWithoutStep(input: { readonly span: Span }): Diagnostic {
  return {
    severity: 'error',
    code: 'error-fence-without-step',
    message:
      'This `error` fence marks the example as expected-to-fail, but the example has no step to run.',
    span: input.span,
  }
}

// A reference block (ADR 0016) points at an oath the workspace does not hold.
// Never prose: a link-only block that resolves to nothing has no other reading,
// so it fails the run rather than degrading silently.
export function referenceNotFound(input: {
  readonly text: string
  readonly path: string
  readonly span: Span
}): Diagnostic {
  return {
    severity: 'error',
    code: 'reference-not-found',
    message:
      `Reference to "${input.text}" points at "${input.path}", which is not an oath in this ` +
      'workspace.\nCheck the path, and that the file is matched by the `docs` globs in ' +
      'varar.config.json.',
    span: input.span,
  }
}

// The referenced document exists but the section contributes no steps — a
// mistyped anchor, or a section that is pure prose.
export function referenceEmpty(input: {
  readonly text: string
  readonly path: string
  readonly slug: string
  readonly span: Span
}): Diagnostic {
  const where = input.slug === '' ? input.path : `${input.path}#${input.slug}`
  return {
    severity: 'error',
    code: 'reference-empty',
    message:
      `Reference to "${input.text}" resolves to "${where}", which contributes no steps.\n` +
      'Check the heading the anchor names, and that its section contains a matching paragraph.',
    span: input.span,
  }
}

// References may nest to any depth (depth is a style question, not a rule), so
// a chain that reaches a section already on it must be reported rather than
// recursed into.
export function referenceCycle(input: {
  readonly chain: ReadonlyArray<string>
  readonly span: Span
}): Diagnostic {
  return {
    severity: 'error',
    code: 'reference-cycle',
    message: `Reference cycle: ${input.chain.join(' → ')}.`,
    span: input.span,
  }
}
